servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ResolvingConnectionFactoryFilter.java

```diff
@@ -0,0 +1,101 @@
+import io.servicetalk.transport.api.HostAndPort;
+import io.servicetalk.transport.api.TransportObserver;
+
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.client.api.ServiceDiscovererEvent.Status.AVAILABLE;
+import static io.servicetalk.http.netty.GlobalDnsServiceDiscoverer.globalDnsServiceDiscoverer;
+
+/**
+ * A {@link ConnectionFactoryFilter} that will resolve the passed unresolved {@link InetSocketAddress} on each attempt
+ * to create a {@link ConnectionFactory#newConnection(Object, ContextMap, TransportObserver) new connection} using
+ * {@link GlobalDnsServiceDiscoverer#globalDnsServiceDiscoverer()}.
+ */
+final class ResolvingConnectionFactoryFilter
+        implements ConnectionFactoryFilter<InetSocketAddress, FilterableStreamingHttpConnection> {
+
+    static final ConnectionFactoryFilter<InetSocketAddress, FilterableStreamingHttpConnection> INSTANCE =
+            new ResolvingConnectionFactoryFilter();
+
+    private final ServiceDiscoverer<HostAndPort, InetSocketAddress, ServiceDiscovererEvent<InetSocketAddress>> sd =
+            globalDnsServiceDiscoverer();
+
+    private ResolvingConnectionFactoryFilter() {
+        // Singleton
+    }
+
+    @Override
+    public ConnectionFactory<InetSocketAddress, FilterableStreamingHttpConnection> create(
+            final ConnectionFactory<InetSocketAddress, FilterableStreamingHttpConnection> original) {
+        return new DelegatingConnectionFactory<InetSocketAddress, FilterableStreamingHttpConnection>(original) {
+
+            @Override
+            public Single<FilterableStreamingHttpConnection> newConnection(final InetSocketAddress address,
+                                                                           @Nullable final ContextMap context,
+                                                                           @Nullable final TransportObserver observer) {
+                assert address.isUnresolved();
+                return sd.discover(HostAndPort.of(address)).takeAtMost(1).firstOrError()
+                        .flatMap(resolvedAddresses -> {
+                            if (resolvedAddresses.size() > 1) {
+                                // In case DNS server returns multiple IPs, it's recommended to shuffle the result to
+                                // make sure the client balances load between all available IPs.
+                                List<ServiceDiscovererEvent<InetSocketAddress>> list =
+                                        resolvedAddresses instanceof List ?
+                                                (List<ServiceDiscovererEvent<InetSocketAddress>>) resolvedAddresses :
+                                                new ArrayList<>(resolvedAddresses);
+                                Collections.shuffle(list);
```

Comment:
From reading this I assume shuffling is the best thing we can do (compared to keeping state for a "true" round robin)?

Reply:
Yes, our best effort to somehow distribute where the client connects to.